cluster-autoscaler/cloudprovider/aws/aws_cloud_provider.go

```diff
@@ -21,20 +21,74 @@ import (
 	"regexp"
 	"strings"
 
+	"errors"
 	"k8s.io/contrib/cluster-autoscaler/cloudprovider"
 	"k8s.io/contrib/cluster-autoscaler/config/dynamic"
 	apiv1 "k8s.io/kubernetes/pkg/api/v1"
 )
 
-// AwsCloudProvider implements CloudProvider interface.
-type AwsCloudProvider struct {
+// awsCloudProvider implements CloudProvider interface.
+type awsCloudProvider struct {
 	awsManager *AwsManager
 	asgs       []*Asg
 }
 
+// autoDiscoveringProvider implements CloudProvider interface.
+type autoDiscoveringProvider struct {
+	*awsCloudProvider
+}
+
 // BuildAwsCloudProvider builds CloudProvider implementation for AWS.
-func BuildAwsCloudProvider(awsManager *AwsManager, specs []string) (*AwsCloudProvider, error) {
-	aws := &AwsCloudProvider{
+func BuildAwsCloudProvider(awsManager *AwsManager, discoveryOpts cloudprovider.NodeGroupDiscoveryOptions) (cloudprovider.CloudProvider, error) {
+	if len(discoveryOpts.NodeGroupSpecs) > 0 {
+		return buildStaticallyDiscoveringProvider(awsManager, discoveryOpts.NodeGroupSpecs)
+	}
+	if discoveryOpts.NodeGroupAutoDiscoverySpec != "" {
+		return buildAutoDiscoveringProvider(awsManager, discoveryOpts.NodeGroupAutoDiscoverySpec)
+	}
+	return nil, errors.New("failed to build an aws cloud provider: either --nodes or --node-group-auto-discovery must be specified")
```

Comment:
I would not put the flags here as this part of code is not really aware of how the flags are defined.